pylint/checkers/typecheck.py

```diff
@@ -1241,13 +1245,28 @@ def visit_compare(self, node):
         if op in ['in', 'not in']:
             self._check_membership_test(right)
 
-    @check_messages('unsubscriptable-object', 'unsupported-assignment-operation',
-                    'unsupported-delete-operation')
+    @check_messages('unsubscriptable-object',
+                    'unsupported-assignment-operation',
+                    'unsupported-delete-operation',
+                    'unhashable-dict-key')
     def visit_subscript(self, node):
         supported_protocol = None
         if isinstance(node.value, (astroid.ListComp, astroid.DictComp)):
             return
 
+        if isinstance(node.value, astroid.Dict):
+            # Assert dict key is hashable
+            try:
+                hash_fn = next(next(
```

Comment:
We have a function called `safe_infer` which you can use. You can find it in `pylint.checkers.utils`. This function returns either `None` or `astroid.Uninferable` so make sure to filter these values. Basically this can look as:
```
elem = safe_infer(node.slice.value)
if elem is None or elem is astroid.Uninferable:
     # don't emit the check
if isinstance(elem, astroid.Instance): # only do this for Instances and/or maybe classes
   # now do the getattr check